### ISSUE TITLE
assertWarns doesn't handle duplicate warnings properly

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2904]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#2904|
|Type|defect|
|Created|2007-11-16 16:23:08Z|

If you have a function, say a recursive function, that has a warning and that you are running with `assertWarns`, then that warning will be recorded multiple times even though the warning is only generated by one line of code.

The error you get is big and ridiculous and unhelpful too.

<details><summary>Searchable metadata</summary>

```
trac-id__2904 2904
type__defect defect
reporter__jml jml
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1195230188000000 1195230188000000
changetime__1207125606000000 1207125606000000
version__None None
owner__therve therve

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [23156]) Fix the assertWarns problem with multiple warnings, to be able to remove the
awkward code in SSHFactory.

Refs [#2904](https://github.com/twisted/twisted/issues/2904)
Refs [#3141](https://github.com/twisted/twisted/issues/3141)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set status to `closed`|
|-|-|

(In [23170]) Merge sshfactory-keys-3141

Author: therve
Reviewer: exarkun
Fixes [#3141](https://github.com/twisted/twisted/issues/3141)
Fixes [#2904](https://github.com/twisted/twisted/issues/2904)

Fix backward compatibility behavior of SSHFactory, allowing the publicKeys and
privateKeys attributes to be string instead of key objects. Fix also a problem
in assertWarns with multiple warnings.
